### PR TITLE
Add new checker `bad-chained-comparison`

### LINKED ISSUE #6559: Warn about ``if x is None == y is None``
### Current problem

We sometimes see code like this:

```python
def foo(*, left=None, right=None):
  """ ... """
  if (left is None) != (right is None):
    raise ValueError('Either both left= and right= need to be provided or none should.')
```

It is very easy to make the mistake of writing the check as:

```python
def foo(*, left=None, right=None):
  """ ... """
  if left is None != right is None:
    raise ValueError('Either both left= and right= need to be provided or none should.')
```
This actually is a chained comparison: ``(left is None != right is None)`` is equivalent to:

```python
(left is None) and (None != right) and (right is None)
```

...which is unsatisfiable, since `right is None` implies `not(None != right)`.

### Desired solution

According to the `comparison` rule in the Python grammar (https://docs.python.org/3/reference/grammar.html),
these comparison operators have the same precedence, and would lead to chained comparisons:

``in, not in, is, is not, <, <=, >, >=, !=, ==``

There are three groups: ``{'in', 'not in'}``, ``{'is', 'is not'}``, ``{'<', '<=', '>', '>=', '!=', '=='}``.

If the linter warned about chained comparisons where one expression is part of two comparisons that belong to different groups, this would flag up checks such as "left is None != right is None".

The price to pay is that this would also trigger on code such as...:

```python
if None is not f(x) > 0:
  ...
```

...but overall, it might be justified in those rare cases where a conditional just like that is appropriate to expect the author to silence the linter on that line.

### PR DESCRIPTION
**Pull request by zenlyj**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
|    | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
|    | :hammer: Refactoring   |
| ✓   | :scroll: Docs          |

## Description

As suggested by the issue, a default checker is added to warn users about potential logical mistakes from a chained comparison consisting of comparisons that belong to different semantic categories/groups. Please take a look at the referenced issue as this implementation is pretty much a 1-1 of the solution suggested there.

Any advice is greatly appreciated!

Closes #6559

### CODE REVIEW

**Comment by Pierre-Sassoulas** [comment]
The new message in primer seems genuine ! Amazing !

File: pylint/extensions/bad_chained_comparison.py
Comment (Pierre-Sassoulas) [inline pylint/extensions/bad_chained_comparison.py:38]:
```suggestion
        return any(o not in group for o in operators)
```
Look like a false negative of the for-any-all checker.

File: pylint/extensions/bad_chained_comparison.py
Comment (Pierre-Sassoulas) [inline pylint/extensions/bad_chained_comparison.py:43]:
```suggestion
            self.add_message("bad-chained-comparison", node=node, args=(num_pieces), confidence=HIGH)

```
As we're not using inference during the check. Applying this change will require and import and regenerating the functional output.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
This looks pretty good already, thank you !

**Comment by nickdrozd** [comment]
I think it would be reasonable to enable this by default and not make it an extension. The one primer result appears to be a full-on error and my guess is that most hits would be also be errors.

I know we try to avoid false positives, but on the other hand, surfacing latent errors helps to highlight the value of the tool.

File: pylint/extensions/bad_chained_comparison.py
Comment (zenlyj) [reply pylint/extensions/bad_chained_comparison.py:38]: I think both options should have the same performance?

Let's say `o not in group for o in operators` evaluates to the following iterable: `[False, True, True]`.
Then, `o in group for o in operators` evaluates to the inverse: `[True, False, False]`.

`any([False, True, True])` short circuits at the first True, and conversely, `all([True, False, False])`  short circuits at the first False.

My point is, I don't think the frequency of operators that are not in the group matters, since the short circuit effect should always occur on the same index in both iterables.

Please correct me if I am missing something here.

**Comment by zenlyj** [comment]
> I think it would be reasonable to enable this by default and not make it an extension. The one primer result appears to be a full-on error and my guess is that most hits would be also be errors.
> 
> I know we try to avoid false positives, but on the other hand, surfacing latent errors helps to highlight the value of the tool.

I agree, I feel that most hits would most likely be errors. I don't think there are many scenarios that warrant such confusing chained comparisons. Any thoughts on this? @Pierre-Sassoulas

File: pylint/extensions/bad_chained_comparison.py
Comment (zenlyj) [reply pylint/extensions/bad_chained_comparison.py]: Done!

File: pylint/extensions/bad_chained_comparison.py
Comment (zenlyj) [reply pylint/extensions/bad_chained_comparison.py]: Refactored according to pylint's suggestions.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
I agree with @nickdrozd, as evidenced by the primer result this is a great check. It's not that often that we discover a real world bug in the primers with a 0% false positive rate from the start. In fact it never happened before. If we want to make this a default check we just have to move it to ``pylint/checker``, please keep the code independent :)

I suggested some wording change and a shorter message with more details, let me know what you think. (I did not review the logic yet)

File: pylint/extensions/bad_chained_comparison.py
Comment (Pierre-Sassoulas) [reply pylint/extensions/bad_chained_comparison.py]: You're right it's identical (Actually I initially made the same mistake in https://github.com/PyCQA/pylint/issues/4146). Resolving, sorry about the noise 😄

File: doc/whatsnew/fragments/6559.new_check
Comment (zenlyj) [reply doc/whatsnew/fragments/6559.new_check]: `semantically incompatible with the other.` sounds good to me, updated!

File: pylint/extensions/bad_chained_comparison.py
Comment (zenlyj) [reply pylint/extensions/bad_chained_comparison.py]: This message should be easier to understand, updated!

File: pylint/extensions/bad_chained_comparison.py
Comment (zenlyj) [reply pylint/extensions/bad_chained_comparison.py]: Updated according to suggestion, made some minor grammar corrections at:
`mean the same thing as "is"` and `chaining them in`

File: tests/functional/ext/bad_chained_comparison/bad_chained_comparison.txt
Comment (zenlyj) [reply tests/functional/ext/bad_chained_comparison/bad_chained_comparison.txt]: Done, message should emit the unique operators (ordered by left to right in expression) used in the chained comparison now.

File: pylint/checkers/bad_chained_comparison.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/bad_chained_comparison.py:27]:
```suggestion
            "Used when there is a chained comparison where one expression is part "
            "of two comparisons that belong to different semantic groups "
            ' ("<" does not mean the same thing as "is", chaining them in '
            '"0 < x is None" is probably a mistake).',

```

File: pylint/checkers/bad_chained_comparison.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/bad_chained_comparison.py:43]:
```suggestion
        operators = set(op[0] for op in node.ops)
```

File: pylint/checkers/bad_chained_comparison.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/bad_chained_comparison.py:33]:
```suggestion
```
We're trying to get a list of uniq operator right ? Shouldn't this be a set as in ``set(op[0] for op in node.ops)`` ?

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: pylint/checkers/bad_chained_comparison.py
Comment (zenlyj) [reply pylint/checkers/bad_chained_comparison.py:33]: Using `set(...)` gives us a list of unique operators, however, the operators will be unordered (from my understanding, python sets are unordered). Therefore, the message output will be non-deterministic.

The `list(dict.fromkeys(...))` gives a list of unique operators as well, and will order the operators according to their position in the expression https://stackoverflow.com/questions/39755464/python-unique-items-of-list-in-the-order-that-it-appears. 
For example: given the expression `(a < b <= c in d)`, the list will always be `['<', '<=', 'in']`. Whereas the set may create `{'<=', 'in', '<'}` in an instance, and output `{'in', '<', '<='}` in another instance.

In my opinion, it would be better if the message output is deterministic. Also, so that we can perform functional testing 🙂 Let me know what you think.

File: pylint/checkers/bad_chained_comparison.py
Comment (zenlyj) [reply pylint/checkers/bad_chained_comparison.py]: updated!

File: pylint/checkers/bad_chained_comparison.py
Comment (Pierre-Sassoulas) [reply pylint/checkers/bad_chained_comparison.py:35]: Maybe we can sort explicitly then ?

File: pylint/checkers/bad_chained_comparison.py
Comment (Pierre-Sassoulas) [reply pylint/checkers/bad_chained_comparison.py:45]:
```suggestion
        operators = sorted(set(op[0] for op in node.ops))

```

File: pylint/checkers/bad_chained_comparison.py
Comment (zenlyj) [reply pylint/checkers/bad_chained_comparison.py:35]: that works too, since the operator set is quite small, sorting shouldn't be too costly. let's go with sort so that the code will be easier to understand.

**Review by Pierre-Sassoulas — DISMISSED** [review]

**Comment by Pierre-Sassoulas** [comment]
I think it's a pretty nice checker, great job @zenlyj. Let's wait for another maintainer review, this is a default check after all, mistakes would affect a lot of users.

File: pylint/checkers/bad_chained_comparison.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/bad_chained_comparison.py:1]:
```suggestion
# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
# Copyright (c) https://github.com/PyCQA/pylint/blob/main/CONTRIBUTORS.txt

from __future__ import annotations
```

File: pylint/checkers/bad_chained_comparison.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/bad_chained_comparison.py:33]:
```suggestion
    def _has_diff_semantic_groups(self, operators: list[str]) -> bool:

```

**Comment by Pierre-Sassoulas** [comment]
I have a really hard time to understand the pre-commit error. I can reproduce locally but I can't seem to fix it.

**Comment by zenlyj** [comment]
@Pierre-Sassoulas it seems that the pre-commit error is due to the files having Windows styles line breaks. It should be fixed now, sorry for the trouble!

Regarding the documentation check fail, is this expected for new documentation pages added?

**Comment by Pierre-Sassoulas** [comment]
Strange, I fixed the line ending locally and it did not  work.

**Comment by Pierre-Sassoulas** [comment]
#8007 has been merged, could you rebase on it please ?

**Comment by Pierre-Sassoulas** [comment]
The documentation check is expected to fail as long as it's not merged in main and in the official doc.

**Review by jacobtylerwalls — APPROVED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]
First check that not only did not have a single primer false positives but even detected a real bug to fix in a very popular well scrutinized project.  It already detected and [permitted to fix a bug in sentry](https://github.com/getsentry/sentry/pull/42669), which is pretty impressive ! Great checker, amazing job 👌

**Pull request merged**
